src/librustc/ty/relate.rs

```diff
@@ -575,6 +585,70 @@ pub fn super_relate_tys<'a, 'gcx, 'tcx, R>(relation: &mut R,
     }
 }
 
+/// The main "const relation" routine. Note that this does not handle
+/// inference artifacts, so you should filter those out before calling
+/// it.
+pub fn super_relate_consts<'a, 'gcx, 'tcx, R>(
+    relation: &mut R,
+    a: &'tcx ty::LazyConst<'tcx>,
+    b: &'tcx ty::LazyConst<'tcx>
+) -> RelateResult<'tcx, &'tcx ty::LazyConst<'tcx>>
+where
+    R: TypeRelation<'a, 'gcx, 'tcx>, 'gcx: 'a+'tcx, 'tcx: 'a
+{
+    let tcx = relation.tcx();
+
+    match (a, b) {
+        (ty::LazyConst::Evaluated(a_eval), ty::LazyConst::Evaluated(b_eval)) => {
+            // Only consts whose types are equal should be compared.
+            assert_eq!(a_eval.ty, b_eval.ty);
+
+            // Currently, the values that can be unified are those that
+            // implement both `PartialEq` and `Eq`, corresponding to
+            // `structural_match` types.
+            // FIXME(const_generics): check for `structural_match` synthetic attribute.
+            match (a_eval.val, b_eval.val) {
+                (ConstValue::Infer(_), _) | (_, ConstValue::Infer(_)) => {
+                    // The caller should handle these cases!
+                    bug!("var types encountered in super_relate_consts: {:?} {:?}", a, b)
+                }
+                (ConstValue::Param(a_p), ConstValue::Param(b_p)) if a_p.index == b_p.index => {
+                    Ok(a)
+                }
+                (ConstValue::Scalar(Scalar::Bits { .. }), _) if a == b => {
+                    Ok(a)
+                }
+                (ConstValue::ByRef(..), _) => {
+                    bug!(
+                        "non-Scalar ConstValue encountered in super_relate_consts {:?} {:?}",
+                        a,
+                        b,
+                    );
+                }
+                 _ => {
+                    Err(TypeError::ConstError(
+                        ConstError::Mismatch(expected_found(relation, &a, &b))
+                    ))
+                }
+            }
+        }
+        // FIXME(const_generics): this is probably wrong (regarding TyProjection)
```

Comment:
Well, to be quite honest, I'm a bit confused by what the "normalization policy" is here. If this were doing eager normalization, as I would expect at the moment, then this doesn't seem wrong. But if it's lazy normalization, then indeed it is wrong. It seems like we can land it as is, but I think that overall it would be good to try and document and talk out the design of constants and const generics.

Reply:
Design meeting! :)